Fix TEMP debugging

TylerLeonhardt opened the pull request:
This isn't quite ready yet. But I wanted to share what it took.

TylerLeonhardt commented on src/PowerShellEditorServices/Services/DebugAdapter/Handlers/ConfigurationDoneHandler.cs:
```diff
@@ -43,10 +43,22 @@ internal class ConfigurationDoneHandler : IConfigurationDoneHandler
+                    await Task.Delay(2000);
+                }
+            }
+
```
@SeeminglyScience

SeeminglyScience replied:
`SessionState` should have a property-changed-like event. If `SessionState` starts at `Ready`, we should add another value to the enum `NotStarted` or something.  That would allow skipping the first delay.

Also maybe just have a `ManualResetEventSlim` that this response can `Wait` on.

TylerLeonhardt replied:
Maybe we put this to `NotStarted` https://github.com/PowerShell/PowerShellEditorServices/blob/master/src/PowerShellEditorServices/Services/PowerShellContext/PowerShellContextService.cs#L371

TylerLeonhardt commented:
NOTE: This last commit requires a change in PSRL.

TylerLeonhardt commented:
Marking as ready for review but it does need that PSRL change to go in.

rjmholt commented on src/PowerShellEditorServices/Hosting/EditorServicesHost.cs:
```diff
@@ -291,6 +292,25 @@ public void StartLogging(string logFilePath, PsesLogLevel logLevel)
             _logger.LogInformation($"Debug NamedPipe: {config.InOutPipeName}\nDebug OutPipe: {config.OutPipeName}");
 
+            IServiceProvider serviceProvider = null;
+            if (!useExistingSession)
```
Is there a reason for not using `useTempSession` with inverted logic? That would seem to line up with the conceptual feature better to me, and feels like it aligns with "breaking out" of the normal code when the `useTempSession` switch is set.

TylerLeonhardt replied:
Yeah that's fine - `useExistingSession` was just "what was there before"

TylerLeonhardt replied:
changed it to `useTempSession` hopefully that's a bit more clear.

rjmholt commented on src/PowerShellEditorServices/Hosting/EditorServicesHost.cs:
```diff
@@ -378,7 +389,12 @@ public void StopServices()
+            if (_languageServer != null)
+            {
+                _languageServer.WaitForShutdown().Wait();
+                return;
```
If there's both a languageServer and a debugServer, why do we only wait for one? Possible to document? Just from reading this code block there's no obvious link to the temp session stuff, but I'm guessing there is and we should make sure that's clear

TylerLeonhardt replied:
Yeah I'll throw a comment. It _is_ related to temp debugging because in temp debugging there is not language server... so we need to wait on something - aka the debug. And in the other scenario, if the language server stops, we wanna stop the whole process

TylerLeonhardt replied:
added.

rjmholt commented on src/PowerShellEditorServices/Server/PsesDebugServer.cs:
```diff
@@ -34,9 +36,10 @@ public class PsesDebugServer : IDisposable
             _loggerFactory = factory;
             _inputStream = inputStream;
             _outputStream = outputStream;
+            _serverStart = new TaskCompletionSource<bool>();
```
```suggestion
            _serverStarted = new TaskCompletionSource<bool>();
```

TylerLeonhardt replied:
gonna change this to `_serverRunning` as it will be used to "wait until the server shuts down"

TylerLeonhardt replied:
changed it to that because when the "result is set" it's like saying set the result of the "running". Let me know if you feel stronger about `_serverStopped`

rjmholt commented on src/PowerShellEditorServices/Server/PsesDebugServer.cs:
```diff
@@ -95,6 +97,12 @@ public void Dispose()
+
+        public async Task WaitForShutdown()
+        {
+            await _serverStart.Task;
```
If this is server start, why does it get set on shutdown? Should it be something like `serverStopped` instead?

TylerLeonhardt replied:
Yeah I guess it should be called like `_serverRunning` or something.

TylerLeonhardt replied:
changed it to that because when the "result is set" it's like saying set the result of the "running". Let me know if you feel stronger about `_serverStopped`

TylerLeonhardt replied:
Changed to `_serverStopped`!

rjmholt commented on src/PowerShellEditorServices/Server/PsesServiceCollection.cs:
```diff
@@ -0,0 +1,85 @@
+
+namespace Microsoft.PowerShell.EditorServices.Server
+{
+    internal static class PsesServiceCollection
```
Given that this is a static extensions class, maybe call it something like `PsesServiceCollectionExtensions` instead (i.e. it doesn't actually implement an collection of services)

TylerLeonhardt commented:
build might fail for this because of needed changes in https://github.com/PowerShell/PowerShellEditorServices/pull/1066

but I've addressed all your feedback @rjmholt

SydneyhSmith commented:
When we merge this change in let's also make sure that we make any necessary changes to our [troubleshooting guide](https://github.com/PowerShell/vscode-powershell/blob/master/docs/troubleshooting.md) and our issue templates

TylerLeonhardt commented on src/PowerShellEditorServices/Hosting/EditorServicesHost.cs:
```diff
@@ -378,7 +384,12 @@ public void StopServices()
-            _languageServer.WaitForShutdown().Wait();
+            if (_languageServer != null)
+            {
+                _languageServer.WaitForShutdown().Wait();
```
fixed!

TylerLeonhardt commented on src/PowerShellEditorServices/Server/PsesServiceCollection.cs:
```diff
@@ -0,0 +1,85 @@
+                    });
+        }
+
+        public static IServiceCollection AddPsesDebugServices (
```
fixed!

TylerLeonhardt commented on src/PowerShellEditorServices/Server/PsesServiceCollection.cs:
```diff
@@ -0,0 +1,85 @@
+            IServiceProvider languageServiceProvider,
+            PsesDebugServer psesDebugServer,
+            bool useExistingSession
+        )
```
Fixed!

TylerLeonhardt commented:
All feedback addressed and rebased! The can be merged in really at any time it just won't work until PSRL is released.

TylerLeonhardt commented:
This is not being backported.
